include/drivers/i2c.h

```diff
@@ -252,6 +255,31 @@ static inline int z_impl_i2c_transfer(struct device *dev,
 	return api->transfer(dev, msgs, num_msgs, addr);
 }
 
+/**
+ * @brief Recover the I2C bus
+ *
+ * Attempt to recover the I2C bus.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @retval 0 If successful
+ * @retval -EBUSY If bus is not clear after recovery attempt.
+ * @retval -EIO General input / output error.
```

Comment:
I think this is a mis-use of `-EBUSY`, which would be appropriate only if the attempt failed because the bus was doing something else.  I'm not sure if it's worth distinguishing "something went wrong during recovery" from "recovery was completed but the bus is still stuck".  Can we just return `-EIO` if the bus is not cleared when the call returns?

Reply:
@pabigot Thank you for reviewing.

If the bus is not clear after the recovery attempt it signals just that; the bus is busy doing something else (not the controller, but the bus itself). I would like to be able to distinguish between a failed IO request and a non-clear bus.

The Linux kernel uses -EBUSY in this regard, too: https://elixir.bootlin.com/linux/latest/source/drivers/i2c/i2c-core-base.c#L207

Let's discuss this in further detail at the API meeting. I will implement needed changes after that.

Reply:
But Linux returns `-EBUSY` only if `SCL` is high at a point where the reset process expects it to be low, meaning something else is clocking the bus.  That's not being checked by your implementation as far as I can see; you're looking at `SDA` at the end, which is a different failure that I don't see Linux checking for.

If we do go with `-EBUSY` it should match the Linux behavior and be documented as "The recovery process was not performed because the bus was busy."  It should not imply anything about whether the bus is clear.

Yes, can discuss at API.

Reply:
Linux is also checking SDA at the very end, see ` i2c_generic_bus_free()` at https://elixir.bootlin.com/linux/latest/source/drivers/i2c/i2c-core-base.c#L153.

Reply:
OK; I don't like this interpretation, but I'll accept it as following prior art.